core/src/muxing.rs

```diff
@@ -51,178 +50,62 @@
 //! The upgrade process will take ownership of the connection, which makes it possible for the
 //! implementation of `StreamMuxer` to control everything that happens on the wire.
 
-use fnv::FnvHashMap;
-use futures::{future, prelude::*, task::Context, task::Poll};
+use futures::{task::Context, task::Poll};
 use multiaddr::Multiaddr;
-use parking_lot::Mutex;
-use std::{
-    fmt, io,
-    ops::Deref,
-    pin::Pin,
-    sync::atomic::{AtomicUsize, Ordering},
-};
+use std::collections::VecDeque;
+use std::io;
 
+pub use self::boxed::StreamMuxerBox;
+pub use self::boxed::SubstreamBox;
 pub use self::singleton::SingletonMuxer;
 
+mod boxed;
 mod singleton;
 
-/// Implemented on objects that can open and manage substreams.
-///
-/// The state of a muxer, as exposed by this API, is the following:
-///
-/// - A connection to the remote. The `poll_event`, `flush_all` and `close` methods operate
-///   on this.
-/// - A list of substreams that are open. The `poll_outbound`, `read_substream`, `write_substream`,
-///   `flush_substream`, `shutdown_substream` and `destroy_substream` methods allow controlling
-///   these entries.
-/// - A list of outbound substreams being opened. The `open_outbound`, `poll_outbound` and
-///   `destroy_outbound` methods allow controlling these entries.
+/// Provides multiplexing for a connection by allowing users to open substreams.
 ///
+/// A [`StreamMuxer`] is a stateful object and needs to be continuously polled to make progress.
 pub trait StreamMuxer {
-    /// Type of the object that represents the raw substream where data can be read and written.
```

Comment:
Comment is roughly accurate, no?

Reply:
Correct. Removing this is a left-over from first removing the associated type and reintroducing it based on the review feedback :D

I'll restore the doc string as well!